dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,442 @@
+            Block bb, Analysis<A, S, T> analysis, String escapeCharacter) {
+        assert analysis != null
+                : "analysis should be non-null when visualizing the transfer input of a block.";
+
+        TransferInput<A, S> input = analysis.getInput(bb);
+        assert input != null;
+
+        StringBuilder sbStore = new StringBuilder();
+
+        // split input representation to two lines
+        sbStore.append("Before: ");
+        if (!input.containsTwoStores()) {
+            S regularStore = input.getRegularStore();
+            sbStore.append(visualizeStore(regularStore));
+        } else {
+            S thenStore = input.getThenStore();
+            sbStore.append("then=");
+            sbStore.append(visualizeStore(thenStore));
+            S elseStore = input.getElseStore();
+            sbStore.append(", else=");
+            sbStore.append(visualizeStore(elseStore));
+        }
+        sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+        return sbStore.toString();
+    }
+
+    /**
+     * Helper method to simplify visualizing a special block; it is useful when implementing a
+     * custom CFGVisualizer.
+     *
+     * @param sbb the special block
+     * @param separator the separator String to use
+     * @return the String representation of the special block
+     */
+    protected String visualizeSpecialBlockHelper(SpecialBlock sbb, String separator) {
+        String specialBlock = "";
+        switch (sbb.getSpecialType()) {
+            case ENTRY:
+                specialBlock = "<entry>" + separator;
+                break;
+            case EXIT:
+                specialBlock = "<exit>" + separator;
+                break;
+            case EXCEPTIONAL_EXIT:
+                specialBlock = "<exceptional-exit>" + separator;
+                break;
+        }
+        return specialBlock;
+    }
+
+    /**
+     * Helper method called by {@link #visualizeBlockHelper}. If possible, get the last node of a
+     * block.
+     *
+     * @param bb the block
+     * @return the last node of this block or {@code null}
+     */
+    protected Node getLastNode(Block bb) {
+        Node lastNode;
+        switch (bb.getType()) {
+            case REGULAR_BLOCK:
+                List<Node> blockContents = ((RegularBlock) bb).getContents();
+                lastNode = blockContents.get(blockContents.size() - 1);
+                break;
+            case EXCEPTION_BLOCK:
+                lastNode = ((ExceptionBlock) bb).getNode();
+                break;
+            default:
+                lastNode = null;
+        }
+        return lastNode;
+    }
+
+    /**
+     * Generate the order of processing blocks.
+     *
+     * @param cfg the current control flow graph
+     * @return the IdentityHashMap which maps from blocks to their orders
+     */
+    protected IdentityHashMap<Block, List<Integer>> getProcessOrder(ControlFlowGraph cfg) {
```

Comment:
Why is the mapped-to value a list?  Is it always a singleton list?

Reply:
`cfg.getDepthFirstOrderedBlocks()` will be called in method `getProcessOrder`, it will return the set of basic blocks in the control flow graph. But, blocks may appear more than once in the set (it is documented in `getDepthFirstOrderedBlocks()`), so I think that's why we use a list instead of a integer.

Reply:
Clients won't read the documentation of `getDepthFirstOrderedBlocks()`.  (They won't even know to do so!)  Please make the documentation of the method stand on its own.

For example, if the list is not a singleton, is the block processed multiple times?  Does it get visualized multiple times even though it exists only once in the graph?